### PR TITLE
Space - Diana

### PR DESCRIPTION
**Pull request by dnguye2**
# Assignment Submission: Media Ranker

Congratulations! You're submitting your assignment. Please reflect on the assignment with these questions.

## Reflection

Prompt | Response
--- | ---
What was a custom model method you wrote? What was it responsible for doing? | One of the custom methods I wrote was `top_ten(category)` in the `Work` model which calculates the top ten works based on the `counter_cache`, `votes_count`, for a given category. Tiebreakers are resolved by sorting alphabetically.
Describe how you approached testing that model method. What edge cases did you come up with? | To test it, I first used fixtures to create a list of works. <br> My fixtures consisted of: <ul><li>11 albums</li><li>3 movies</li></ul> From the `works` fixture, I made an array of 10 albums I wanted as a top ten (`top_ten_albums_test`) and created a variable `number_of_upvotes` with an initial value of 20. Then, I iterated through the array using an `each` loop. With each iteration, I used a test helper method I created, `perform_upvote`, which would upvote the work `number_of_upvotes` times. And outside of the `times` loop, `number_of_upvotes` was decremented by 2. Testing my method should match the order from the `top_ten_albums_test`. <br><br> Edge cases: <ul><li>tie</li><li>no works</li><li>if there are less than ten items in a category</li></ul>
What are `session` and `flash`? What is the difference between them? | `Session` is used to store info about a user when they login, whereas `flash` denotes a one time message which is generally a success or failure message. `Session` is available until a user closes the browser and `flash` is available from wither the rest of a request cycle until the next complete request or just for one request.
What was one thing that you gained more clarity on through this assignment? | The top thing I gained clarity on was using fixtures to DRY up writing tests.
What is the Heroku URL of your deployed application? | http://retro-media-ranker.herokuapp.com/

### CODE REVIEW

File: app/assets/stylesheets/application.scss [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+@import "bootstrap";
+
+@import "**/*";
+
+// Fonts
+@import url('https://fonts.googleapis.com/css2?family=Exo+2:wght@300;800&display=swap');
+
+// For all content
+body {
+  /* to aid in making the footer sticky */
```
Comment (kaidamasaki) [inline app/assets/stylesheets/application.scss:25]: It's probably simpler to make the footer sticky by just using a `grid`.

File: app/controllers/works_controller.rb [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    @books = Work.sorted_media("book")
+    @movies = Work.sorted_media("movie")
+  end
+
+  def show
+    if @work.nil?
+      flash[:error] = "Work has either been deleted or not found."
+      redirect_to root_path
+      return
+    end
```
Comment (kaidamasaki) [inline app/controllers/works_controller.rb:16]: You should move this sort of logic into `find_work` since that way you get this nice behavior everywhere for "free".

Remember, in a controller filter if it produces a response by doing something like `redirect_to` the action will immediately return to the user without running the action's method.

File: app/models/vote.rb [hunk 3]
```diff
@@ -0,0 +1,9 @@
+class Vote < ApplicationRecord
+  # Validations
+  validates :user_id, presence: true
+  validates :work_id, presence: true
+
+  # Relationships
+  belongs_to :user, counter_cache: true
+  belongs_to :work, counter_cache: true
```
Comment (kaidamasaki) [inline app/models/vote.rb:8]: Nice use of `counter_cache`!

File: app/models/work.rb [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+  validates :publication_year, presence: true
+  validates :description, presence: true
+
+  # Relationships
+  has_many :votes, dependent: :destroy
+  has_many :users, :through => :votes
+
+  # Methods
+  def self.spotlight
+    return spotlight_media = Work.order("votes_count DESC").first
```
Comment (kaidamasaki) [inline app/models/work.rb:16]: It's a little cleaner to use the keyword arguments instead of a string here here:

```suggestion
    return spotlight_media = Work.order(votes_count: :desc).first
```

File: app/models/work.rb [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+  end
+
+  def self.top_ten(category)
+    return nil if (Work.all).empty?
+
+    top_ten_list = Work.where(category: category).sort_by{|work| [-work.votes_count, work.title]}.first(10)
```
Comment (kaidamasaki) [inline app/models/work.rb:22]: You can use `order` here as well:

```suggestion
    top_ten_list = Work.where(category: category).order(votes_count: :desc, title: :asc).first(10)
```

File: app/views/layouts/application.html.erb [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+      <%# General Header %>
+      <div class="container header-custom">
+        <div class="blog-header py-3">
+          <div class="row flex-nowrap justify-content-between align-items-center text-center">
+            <div class="container-fluid text-center">
+              <%= link_to image_tag("https://image.flaticon.com/icons/svg/2540/2540678.svg", height: "60", width: "60", :alt => "cassette tape logo", :class => "d-inline-block align-top"), root_path %>
+              <%= link_to "MEDIA RANKER", root_path, :class => "blog-header-logo header-logo__gradient" %>
+              <br>
+            <h6> STYLING ISN'T THE SAME—I KNOW, I DIDN'T READ THE OPTIONAL INSTRUCTIONS BEFOREHAND 🙃</h6>
+            <h6> EST. 2020 | RANKING THE BEST OF EVERYTHING </h6>
```
Comment (kaidamasaki) [inline app/views/layouts/application.html.erb:25]: You shouldn't have used a `<h6>` here just to get the text size you wanted.

File: app/views/layouts/application.html.erb [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+        <%= yield %>
+      </div>
+    </section>
+    
+    <%# Footer %>
+    <footer class="footer text-center">
+      <div class="container mt-4">
+
+          <small>ALL IMAGES AND TEXT &copy; 2020 MEDIA RANKER </small> <br>
+          <small>Icons made by <a href="https://www.flaticon.com/authors/smalllikeart" title="smalllikeart">smalllikeart</a> from <a href="https://www.flaticon.com/" title="Flaticon">www.flaticon.com</a></small>
```
Comment (kaidamasaki) [inline app/views/layouts/application.html.erb:92]: Using `<small>` and `<br>` should generally be avoided if possible since they don't carry inherent semantic meaning.  (`<small>` was kind of retroactively given a meaning but should still be avoided when you can.)

File: app/views/works/index.html.erb [hunk 8]
```diff
@@ -0,0 +1,13 @@
+<% if @works.length > 0 %>
+    <h1 class="text-center">List of works</h1>
+      <%= render partial: "media-table", locals: { collection: @albums } %>
+      <%= render partial: "media-table", locals: { collection: @books } %>
+      <%= render partial: "media-table", locals: { collection: @movies } %>
```
Comment (kaidamasaki) [inline app/views/works/index.html.erb:5]: Awesome usage of partials here!

**Comment by kaidamasaki** [comment]
# Media Ranker





## Functional Requirements: Manual Testing

| Criteria | yes/no |
| --- | --- |
| **Before logging in** | --
| 1. On index page, there are at most 10 pieces of media on three lists, and a Media Spotlight | ✔️
| 2. Can go into a work's show page | ✔️
| 3. Verify unable to vote on a work, and get a flash message | ✔️
| 4. Can edit this work successfully, and get a flash message | ✔️
| 5. Can go to "View all media" page and see three lists of works, sorted by vote | ✔️
| 6. Verify unable to create a new work when the form is empty, and details about the validation errors are visible to the user through a flash message | ✔️
| 7. Can create a new work successfully. Note the URL for this work's show page | ✔️
| 8. Can delete this work successfully |✔️
| 9. Going back to the URL of this deleted work's show page produces a 404 or some redirect behavior (and does not try to produce a broken view) | ✔️
| 10. Verify that the "View all users" page lists no users | ✔️
| **Log in** | --
| 11. Logging in with a valid name changes the UI to "Logged in as" and "Logout" buttons | ✔️
| 12. Your username is listed in "View all users" page | ✔️
| 13. Verify that number of votes determines the Media Spotlight | ✔️
| 14. Voting on several different pieces of media affects the "Votes" tables shown in the work's show page and the user's show page | ✔️
| 15. Voting on the same work twice produces an error and flash message, and there is no extra vote | ✔️
| **Log out** | --
| 16. Logging out showed a flash message and changed the UI | ✔️
| 17. Logging in as a new user creates a new user | ✔️
| 18. Logging in as an already existing user has a specific flash message | ✔️

## Major Learning Goals/Code Review

| Criteria | yes/no |
| --- | --- |
| 1. Sees the full development cycle including deployment, and the app is deployed to Heroku | ✔️
| 2. Practices full-stack development and fulfilling story requirements: the styling, look, and feel of the app is similar to the original Media Ranker | Styling differs.  (But is pretty sweet!)
| 3. Practices git with at least 25 small commits and meaningful commit messages | ✔️

### Previous Rails learning, Building Complex Model Logic, DRYing up Rails Code

| Criteria | yes/no |
| --- | --- |
| 4. Routes file uses `resources` for works | ✔️
| 5. Routes file shows intention in limiting routes for voting, log-in functionality, and users | ✔️
| 6. The homepage view, all media view, and new works view use semantic HTML | ✔️
| 7. The homepage view, all media view, and new works view use partials when appropriate | ✔️
| 8. The model for media (likely named `work.rb`) `has_many` votes | ✔️
| 9. The model for media has methods to describe business logic, specifically for top ten and top media, possibly also for getting works by some category | ✔️
| 10. Some controller, likely the `ApplicationController`, has a controller filter for finding a logged in user | ✔️
| 11. Some controller, likely the `WorksController`, has a controller filter for finding a work | ✔️
| 12. The `WorksController` uses strong params | ✔️
| 13. The `WorksController`'s code style is clean, and focused on working with requests, responses, `params`, `session`, `flash` | ✔️

### Testing Rails Apps

| Criteria | yes/no |
| --- | --- |
| 14. There are valid fixtures files used for users, votes, and works | No `votes.yml`.
| 15. User model has tests with sections on validations (valid and invalid) and relationships (has votes) | ✔️
| 16. Vote model has tests with sections on validations (valid and invalid) and relationships (belongs to a user, belongs to a vote) | ✔️
| 17. Work model has tests with sections on validations (valid and invalid) and relationships (has votes) | ✔️
| 18. Work model has tests with a section on all business logic methods in the model, including their edge cases | ✔️

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 14+ in Functional Requirements: Manual Testing && 14+ in Code Review | ✔️
| Yellow (Approaches Standards) | 12+ in Functional Requirements: Manual Testing && 11+ in Code Review, or the instructor judges that this project needs special attention |
| Red (Not at Standard) | 0-10 in Code Review or 0-11 in Functional Reqs, or assignment is breaking/doesn’t run with less than 5 minutes of debugging, or the instructor judges that this project needs special attention |





## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Perfect Indentation | ✅
| Elegant/Clever | ✅
| Descriptive/Readable | ✅
| Concise | ✅
| Logical/Organized | ✅

**Review by kaidamasaki — COMMENTED** [review]
Excellent job!  There were a few little ways you could clean your code up but otherwise things looked great!  😄
